src/core/opamStd.ml

```diff
@@ -879,21 +878,67 @@ module OpamSys = struct
       f [] 0 "" 0 true
     else fun path ->
       let split = if clean then OpamString.split else OpamString.split_delim in
-      split path path_sep
+      split path sep
+end
+
+module type Runner = sig
+  type 'a t
+
+  val bind : 'a t -> ('a -> 'b t) -> 'b t
+
+  val map : 'a t -> ('a -> 'b) -> 'b t
+
+  val return : 'a -> 'a t
+
+  val with_process_in : string -> string -> string t
+
+  val run : string -> string -> string option t
+
+  val escape : 'a t -> 'a
+end
+
+module UnitRunner : Runner = struct
+  type 'a t = unit -> 'a
 
-  let with_process_in cmd args f =
+  let bind v f = f (v ())
+
+  let return v = (fun () -> v)
+
+  let map v f = return (f (v ()))
+
+  let with_process_in cmd args =
     if Sys.win32 then
       assert false;
-    let path = split_path_variable (Env.get "PATH") in
+    let path = Path.split_variable (Env.get "PATH") in
```

Comment:
It would be great if callers could do the `which` lookups itself and patch absolute paths. Dune has its own mechanism for looking up binaries in PATH which respects the watch mode.

Reply:
I've kept it as-is to not change the current OPAM behavior and API. Dune would use a different runner implementation where we can look up binaries the Dune way.

Reply:
Sounds good. Can the API be tweaked not to require the shell? I notice that currently:

1. The command and arguments are concatenated. Would it be possible to take a `prog:string -> argv:string list` instead?
2. Instead of relying on shell redirection, can we set stderr/stdout/stdin by providing fd's?